### ISSUE TITLE
How to include Open Source Tables in MSL

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#1102]
**Reported by otter on 15 May 2013 10:18 UTC**
We need to discuss how to include the Open Source Tables in the MSL trunk. The most important questions to answer are:

 1. **Include them directly in MSL or indirectly in ModelicaServices?**
If the tables are directly used in Modelica.Blocks, every vendor needs to use this implementation of the tables.
If the tables are instead included in ModelicaServices and then used from Modelica.Blocks, every vendor can provide its own implementation of the tables.
My opinion: Include the tables directly in Modelica.Blocks since simpler and the end-user always gets exactly the same table implementation and there are no surprises when changing the tool. Furthermore, the open source tables add quite a lot of new functionality and in the future via contracts this will be further extended. It seems therefore that all tool vendors profit from using this implementation.
 1. **How to provide the C-Code/Binaries?**
 The tables are basically a small Modelica layer on external objects in form of C-Code. In principal it would be best to also provide DLLs for all relevant targets under Resources since then MSL is self-contained and it could be used directly from any tool. However, it seems that every tool vendor has a different strategy and uses different compilers and compiler options (MingW, VisualC++, DLL, static lib, etc.).
The alternative is that only the C-Code is provided and every tool
vendor is responsible how to compile and link this code to its system.
 1. **Support of which features to read data from file?**
ITI has provided more as asked in the contract. In particular the implementation provides the possibility to read the following file formats:
Matlab V4 (included in C-source; required by contract)
Matlab V6 (included in C-source)
Matlab V7 (the current default save/load; requires third-party zlib)
Matlab V7.3 (optional, HDF5-based solution; requires third-party HDF5)
I would propose to at least include V7, since then standard Matlab files can be read and it is not much overhead (zlib is small).
I would also be nice to already include V7.3 since MA plans to go in the future for the HDF5 format (very powerful and vendor neutral).
However, this might add quite some overhead for tool vendors, since a static library of HDF5 is around 9 Mbyte.
 1. **How to handle new options in a backwards compatible way?**
The new table implementation adds new features. These new features are activated by new entries in existing enumerations:
Smoothness: new element "ConstantSegments"
Extrapolation: new element "NoExtrapolation"
How to handle this in a backwards compatible way. I think in principal it should be possible to just use the extended enumerations and everything should work. 

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1102

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#x2e;se on 15 May 2013 10:32 UTC**
1. We already use Modelica.Blocks, and vendors have their own table implementation in the C-code. So using ModelicaServices does not change anything.
1. We already provide C-code without providing libraries. Why would we change this?
1. The ITI code requires very specific versions of libhdf5 (does not work with most Linux distributions). zlib is always available; there is a good chance the tool already uses it (it is included in both FMIL and OpenModelica). This should not be a concern for the library anyway. Don't you simply specify a file and the library tries to open it? If it fails, you should get an error message. If the functionality was compiled in, it would simply work.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 15 May 2013 10:57 UTC**
Replying to [comment:1 sjoelund.se]:
> 1. The ITI code requires very specific versions of libhdf5 (does not work with most Linux distributions).

Can you please be more specific about the version? I checked the config/make files of the matio project and all it does is to link with -lhdf5. E.g. OpenSUSE 12.3 installs HDF5-1.8.10 that should then work out of the box.

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#x2e;se on 15 May 2013 11:03 UTC**
For example Debian 6.0.7 (released February 23rd, 2013) uses libhdf5 1.8.4. It is not compatible. The headers are different.

Some API functions for the API are not there (hdf5 as you probably know changed their API recently; so now there are compatibility headers between the old and new API); not everything works until you use the very latest versions.

And hdf5 is huge, so you really do not want to compile it all yourself.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 15 May 2013 11:12 UTC**
Replying to [comment:1 sjoelund.se]:


> 1. We already use Modelica.Blocks, and vendors have their own table implementation in the C-code. So using ModelicaServices does not change anything.

The new table blocks have a different implementation and call different C-functions. For example, the new tables blocks are implemented in a "clean" way as external objects, whereas the MSL 3.2 tables are not (since external objects have not been available when the tables have been implemented a long time ago). As a result, when just including the new table blocks into MSL trunk, no tool will work anymore. The simplest to the end-user would be to provide the new Modelica blocks and binaries under Resources at the same time. Then most tools would at once work without having an interaction with a tool vendor.

> 1. We already provide C-code without providing libraries. Why would we change this?



Previously, there was no way to do it differently. We have introduced an automatic mechanism to link binaries and then it seems naturally to use it also (as, e.g., already done for the Modelica_DeviceDrivers library). If this works not yet as needed, the Modelica specification should be improved using the tables as a benchmark. 
For example, assume a bug-fix is done in the C-code, or new features are added. If only C-code is provided, an end-user has always to wait until a vendor supports this change. If MA would provide binaries, an end-user could at once utilizing the new features by just downloading a new version of MSL.

>

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 15 May 2013 11:33 UTC**
Replying to [comment:1 sjoelund.se]:


> 1. The ITI code requires very specific versions of libhdf5 (does not work with most Linux distributions). zlib is always available; there is a good chance the tool already uses it (it is included in both FMIL and OpenModelica). This should not be a concern for the library anyway. Don't you simply specify a file and the library tries to open it? If it fails, you should get an error message. If the functionality was compiled in, it would simply work.

The issue is the documentation: The documentation of the tables in MSL should define/describe the basic functionality that is expected to be supported by all tools. Futhermore, there will be examples and tests in ModelicaTest and then suddenly a model fails, if a tool does not support a file format. Since it seems that zlib is no problem we should state support for Matlab v7 in the documentation and probably leave the hdf5 open to the tool vendor (not including it in the documentation).

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 15 May 2013 11:47 UTC**
Replying to [comment:4 otter]:
> As a result, when just including the new table blocks into MSL trunk, no tool will work anymore.

Actually, OpenModelica will work without user interaction (we already ship the library).

> Previously, there was no way to do it differently. We have introduced an automatic mechanism to link binaries and then it seems naturally to use it also (as, e.g., already done for the Modelica_DeviceDrivers library). If this works not yet as needed, the Modelica specification should be improved using the tables as a benchmark. 
> For example, assume a bug-fix is done in the C-code, or new features are added. If only C-code is provided, an end-user has always to wait until a vendor supports this change. If MA would provide binaries, an end-user could at once utilizing the new features by just downloading a new version of MSL.

The specification might be there, but:
1. You need to ship a ton of different libraries.
1. The specification does not consider that on the same OS and architecture you still need to ship different versions because for example the C-library might be different (you can not create a "64-bit linux" file and expect it to work for all people on 64-bit linux). The same problem exists on Windows with MinGW-generated objects being incompatible with MSVC-generated objects.

You reference Modelica_DeviceDrivers... Which is a library that only uses that type of linking for a few functions. It has code like:
```mo
within Modelica_DeviceDrivers.OperatingSystem;
function randomReal "returns a random real within the given Range."
input Real minValue = 0;
input Real maxValue = 1;
output Real y;
external "C" y = OS_getRandomNumberDouble(minValue, maxValue);
annotation (Include = "

#include <stdlib.h>
#include <time.h>
static int _randomGeneratorInitialized = 0;
double OS_getRandomNumberDouble(double minValue, double maxValue)
{
  int randomInteger;
  double randomDouble;
  if(!_randomGeneratorInitialized)
  {
    srand ( clock(NULL) * time(NULL) );
    _randomGeneratorInitialized = 1;
  }
  randomInteger = rand();
  randomDouble = ((double)randomInteger/(double)RAND_MAX) * (maxValue - minValue) + minValue;
  return randomDouble;
}



");
end randomReal;
```

I think you are better off letting tool vendors compile the correct version for the user. Probably shaves off the 91kB that static libz adds.

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#x2e;se on 15 May 2013 11:50 UTC**
Replying to [comment:5 otter]:
> The issue is the documentation: The documentation of the tables in MSL should define/describe the basic functionality that is expected to be supported by all tools. Futhermore, there will be examples and tests in ModelicaTest and then suddenly a model fails, if a tool does not support a file format. Since it seems that zlib is no problem we should state support for Matlab v7 in the documentation and probably leave the hdf5 open to the tool vendor (not including it in the documentation).

Nothing is preventing the documentation to list optionally supported types.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 15 May 2013 11:52 UTC**
Replying to [comment:3 sjoelund.se]:
> For example Debian 6.0.7 (released February 23rd, 2013) uses libhdf5 1.8.4. It is not compatible. The headers are different.
> 
> Some API functions for the API are not there (hdf5 as you probably know changed their API recently; so now there are compatibility headers between the old and new API); not everything works until you use the very latest versions.
> 
> And hdf5 is huge, so you really do not want to compile it all yourself.

HDF5 API changed with version 1.8. Matio project uses HDF5 API of version 1.8. I downloaded and compiled HDF5-1.8.4-patch1, compiled it as static lib on VS2005 and could successfully link it to ModelicaTables.dll. So matio project requires at least HDF5 1.8 and I do not understand why it fails with your HDF5 1.8.4 on Debian.

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 15 May 2013 12:01 UTC**
Version: 1.8.4-patch1-3ubuntu2:
```
/usr/include/H5Epublic.h:212:15: note: expected 'void **' but argument is of type 'herr_t (**)(void *)'
../Modelica/NewTables/Resources/C-Sources/ModelicaMatIO.c:17820:5: error: too many arguments to function 'H5Eget_auto1'
/usr/include/H5Epublic.h:212:15: note: declared here
../Modelica/NewTables/Resources/C-Sources/ModelicaMatIO.c:17821:5: error: too many arguments to function 'H5Eset_auto1'
/usr/include/H5Epublic.h:216:15: note: declared here
../Modelica/NewTables/Resources/C-Sources/ModelicaMatIO.c:17845:5: error: too many arguments to function 'H5Eset_auto1'
/usr/include/H5Epublic.h:216:15: note: declared here
../Modelica/NewTables/Resources/C-Sources/ModelicaMatIO.c:17890:9: warning: passing argument 2 of 'H5Eget_auto1' from incompatible pointer type [enabled by default]
/usr/include/H5Epublic.h:212:15: note: expected 'void **' but argument is of type 'herr_t (**)(void *)'
../Modelica/NewTables/Resources/C-Sources/ModelicaMatIO.c:17890:9: error: too many arguments to function 'H5Eget_auto1'
...
```
And it goes on and on and on and on and on and ... Don't ask me why the damn headers point to auto1 instead of auto2; I didn't write them ;) I just know out of the 12 linux releases I build packages for, 6 fail (lucid,precise,squeeze), and 6 work (quantal,raring,wheezy).

Precise (Ubuntu 12.04) and its derivatives (like Mint) will be the most commonly used Linux distribution for another year or so.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 15 May 2013 12:27 UTC**
I compared H5Epublic.h from official HDF5-1.8.4-patch1 and HDF5-1.8.11 and could not find any differences.

I have a
```
#define H5Eget_auto H5Eget_auto2
```
in H5version.h of both HDF5 versions. Can you please check your defines!

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#x2e;se on 15 May 2013 13:56 UTC**
I diff'ed the 1.8.4 and 1.8.9 headers. 1.8.4 has H5_USE_16_API_DEFAULT. Which means it will force v1.6 API (with no single #define to change to v1.8 API; I guess this is why Debian devs changed it for later releases).

Using the 1.8 API requires using the "2" functions, knowing that the library was not compiled with H5_USE_16_API_DEFAULT, or adding:
```
#define H5Acreate_vers 2
#define H5Aiterate_vers 2
#define H5Dcreate_vers 2
#define H5Dopen_vers 2
#define H5Eclear_vers 2
#define H5Eprint_vers 2
#define H5Epush_vers 2
#define H5Eset_auto_vers 2
#define H5Eget_auto_vers 2
#define H5Ewalk_vers 2
#define H5Gcreate_vers 2
#define H5Gopen_vers 2
#define H5Pget_filter_vers 2
#define H5Pget_filter_by_id_vers 2
#define H5Pinsert_vers 2
#define H5Pregister_vers 2
#define H5Rget_obj_type_vers 2
#define H5Tarray_create_vers 2
#define H5Tcommit_vers 2
#define H5Tget_array_dims_vers 2
#define H5Topen_vers 2
```
The latter is probably easiest.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 15 May 2013 14:51 UTC**
You may check config of matio to see how they check for correct HDF5 API: http://matio.git.sourceforge.net/git/gitweb.cgi?p=matio/matio;a=blob;f=config/matio_hdf5.m4;h=b4cdb0ffe398065d0b6ae6b750ef21b8d36a370a;hb=HEAD

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 15 May 2013 15:53 UTC**
Making such a check is easy, but... You already preprocessed the libmatio code to assume hdf5 v1.8. I will make a commit that adds the required #define's.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 15 May 2013 16:03 UTC**
You could add this check to your configure.ac.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 15 May 2013 16:04 UTC**
It is the matio itself that requires HDF5 1.8 API: see http://matio.git.sourceforge.net/git/gitweb.cgi?p=matio/matio;a=commit;h=2e9add807e1b1a1f37c9c62dc7d3f050931df06b

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 15 May 2013 19:44 UTC**
Yeah, that matio code would not work... Fixed in r6362 (I think; compiling everything will take 4 hours or so), but I still am unsure if assuming hdf5 being available is a good idea in the documentation.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 15 May 2013 19:54 UTC**
As nobody from OM followed today's web conference and conclusions are not yet published I can already tell you that availability of HDF5 is not required. Doc will be changed that tools support at least v4, v6 and v7. This is a good agreement as zlib is small and generally available and v7 is (still) the default MAT file format of MATLAB. Support v7.3 is then left optional to tool vendors even more once new tables move to ModelicaExternalC.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 16 May 2013 06:34 UTC**
Here is the result of the discussion at the web/phone meeting of the MAP-LIB group yesterday:

Replying to [ticket:1102 otter]:
> We need to discuss how to include the Open Source Tables in the MSL trunk. The most important questions to answer are:
> 
>  1. **Include them directly in MSL or indirectly in ModelicaServices?**
If the tables are directly used in Modelica.Blocks, every vendor needs to use this implementation of the tables.
If the tables are instead included in ModelicaServices and then used from Modelica.Blocks, every vendor can provide its own implementation of the tables.
My opinion: Include the tables directly in Modelica.Blocks since simpler and the end-user always gets exactly the same table implementation and there are no surprises when changing the tool. Furthermore, the open source tables add quite a lot of new functionality and in the future via contracts this will be further extended. It seems therefore that all tool vendors profit from using this implementation.

Summary of discussion:
Copy Table blocks into MSL now and keep a separate binary library during beta testing. Latest before release candidate of MSL 3.2.1, the library will be included in ModelicaExternal. So tool vendors should during the Beta-Testing phase compile the table source code and include it in their ModelicaExternal binary library.


>  1. **How to provide the C-Code/Binaries?**
 The tables are basically a small Modelica layer on external objects in form of C-Code. In principal it would be best to also provide DLLs for all relevant targets under Resources since then MSL is self-contained and it could be used directly from any tool. However, it seems that every tool vendor has a different strategy and uses different compilers and compiler options (MingW, VisualC++, DLL, static lib, etc.).
The alternative is that only the C-Code is provided and every tool
vendor is responsible how to compile and link this code to its system.

Summary:
Vendors are responsible for it and vendors include it in ModelicaExternal. Table C-Code is stored at the same place as ModelicaInternal.c etc.

Which prefix to use for the C-Code. Currently
  ModelicaNewTables_xxx

The current tables have prefix
  ModelicaTables_xxx

Proposal
  ModelicaStandardTables_xxx

Conclusion: Change prefix to ModelicaStandardTables_xxx
Thomas will perform this change.

>  1. **Support of which features to read data from file?**
ITI has provided more as asked in the contract. In particular the implementation provides the possibility to read the following file formats:
Matlab V4 (included in C-source; required by contract)
Matlab V6 (included in C-source)
Matlab V7 (the current default save/load; requires third-party zlib)
Matlab V7.3 (optional, HDF5-based solution; requires third-party HDF5)
I would propose to at least include V7, since then standard Matlab files can be read and it is not much overhead (zlib is small).
I would also be nice to already include V7.3 since MA plans to go in the future for the HDF5 format (very powerful and vendor neutral).
However, this might add quite some overhead for tool vendors, since a static library of HDF5 is around 9 Mbyte.

Martin Sjoelund: Its a vendor issue.
Thomas B.: Its a vendor issue.
Peter A: At least in the documentation it should be stated that it is optional
Thomas B.: No code from MathWorks used in zlib, only the specification of the interface
Peter A: Would be nice to figure out without simulating what features are supported. For example with some constants.

Conclusion: In documentation of the table blocks it will be stated that V4, V6, V7 is supported and have a remark that tools may support more formats.

>  1. **How to handle new options in a backwards compatible way?**
The new table implementation adds new features. These new features are activated by new entries in existing enumerations:
Smoothness: new element "ConstantSegments"
Extrapolation: new element "NoExtrapolation"
How to handle this in a backwards compatible way. I think in principal it should be possible to just use the extended enumerations and everything should work. 

Hans: New enumeration values should be at the end.
Hilding: In Modelica there is a conversion operator from enumeration
to integer and then there might be problems, if the order is changed.

Conclusion: Seems to be o.k. but move new values to the end (in order to be fully backwards compatible).
Thomas B. will do it.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 16 May 2013 13:15 UTC**
In the current documentation of CombiTimeTable the enumeration values of Smoothness and Extrapolation start with zero. I fixed this in r6367 for the new table implementation. Can someone please cross-check if this is correct.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 16 May 2013 14:57 UTC**
My assigned issues of topics 2, 3 and 4 should all be resolved by r6370.

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#x2e;se on 16 May 2013 15:02 UTC**
Replying to [comment:18 otter]:
> Summary of discussion:
> Copy Table blocks into MSL now and keep a separate binary library during beta testing. Latest before release candidate of MSL 3.2.1, the library will be included in ModelicaExternal. So tool vendors should during the Beta-Testing phase compile the table source code and include it in their ModelicaExternal binary library.

So the library should be changed to use `Library="ModelicaExternalC"` as soon as possible, right?

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#x2e;se on 20 May 2013 06:38 UTC**
Replying to [comment:18 otter]:
> Summary:
> Vendors are responsible for it and vendors include it in ModelicaExternal. Table C-Code is stored at the same place as ModelicaInternal.c etc.

Martin, why was the tables library added in 46cd49d62e1f998b6a3ddcbbb74c2c8d8e40ec55 1f5b3741366d365dafe559936782c2a1daed4682? They take up 1.2MB of space on svn (and I guess released zip's also?). If vendors anyway should make it part of ModelicaExternal, why is it compiled here?

And if it's included in ModelicaExternalC, it should not be named `ModelicaStandardTables`, but `ModelicaExternalC`.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 20 May 2013 07:20 UTC**
Replying to [comment:22 sjoelund.se]:

> Replying to [comment:18 otter]:
> 
> 
> 
> > Summary:
> > Vendors are responsible for it and vendors include it in ModelicaExternal. Table C-Code is stored at the same place as ModelicaInternal.c etc.
> > 
> > 
> > 
> 
> Martin, why was the tables library added in 46cd49d62e1f998b6a3ddcbbb74c2c8d8e40ec55 1f5b3741366d365dafe559936782c2a1daed4682? They take up 1.2MB of space on svn (and I guess released zip's also?). If vendors anyway should make it part of ModelicaExternal, why is it compiled here?
> 
> And if it's included in ModelicaExternalC, it should not be named `ModelicaStandardTables`, but `ModelicaExternalC`.

I just finalized the table inclusion in Modelica and ModelicaTest. Especially, I changed the library annotation to {"ModelicaExternalC", "ModelicaStandardTables"}.

The problem is that no one can work with the tables from MSL (and also cannot test it), if he/she does not get a new ModelicaExternalC object library. That is most likely not an issue for a free tool like OpenModelica. However, commercial tools have usually no mechanism that every user can just download from the web a development version. Furthermore, before any tool can provide it, the one who integrates the tables in the trunk (me) should be able to test the integration, but without having a new ModelicaExternalC library.

Based on the discussion at the last web/phone meeting, I think a meaningful approach is the following:

 1. In the trunk the object libraries of the tables are stored under Resources and both ModelicaExternalC and  ModelicaStandardTables are used in the library annotation. If a tool has the tables already in ModelicaExternalC, this will be used. If a tool does not yet have it, it links the library from Resources. Therefore, for some transition phase, people can use the MSL version from trunk.
 1. When releasing a new version (especially also the Beta release on Wednesday), a copy is made under tags. After doing this, the Resources\libraries directory is removed and the ModelicaStandardTables library is removed from the annotation in the tagged version (this holds for the Beta, ReleaseCandidate and final release). As a result, the tagged versions can always be used as they should. Instead, the trunk version can be used even if the tables are not included in ModelicaExternalC. Once the tables are included in ModelicaExternalC by all tools, the Resources\libraries directory is removed from the trunk.

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 20 May 2013 11:36 UTC**
If you really want tables to be separate they should be separate forever or you get problems and weird workarounds. You could anyway simply compile a Dymola or generic win32 version of the modelicaexternalc like you did with tables. And just replace the vendor version. If the vendor did not think of this use case they designed library handling wrong.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 20 May 2013 12:04 UTC**
Replying to [comment:24 sjoelund.se]:


> If you really want tables to be separate they should be separate forever or you get problems and weird workarounds. 

Why? The tagged version is without a workaround. Generating the tagged version is simple: Removing directory Resources\Library and replacing with one editor command the string ",ModelicaStandardTables" by "" in all files of directory Blocks.


> You could anyway simply compile a Dymola or generic win32 version of the modelicaexternalc like you did with tables. And just replace the vendor version. If the vendor did not think of this use case they designed library handling wrong.

This is not practical: I need to provide the ModelicaExternalC.lib library for Dymola at a particular place on the svn and make a description where to store this library. Storing this library requires usually administrator rights. So, every user of MSL on the trunk using Dymola will first ask why are the tables no longer working and one needs to point to the procedure how to fix it and the person needs administrator rights to do it (which people in industry and also at DLR do not have). Furthermore, it might be that Dymola stores also other object files in ModelicaExternalC (which I do not know). In this case something in Dymola will no longer work. Finally, I am pretty sure that the license conditions of all DS products do not allow this (to change any part of a DS product installation, and especially no binary objects; this is standard for every commercial tool). With the current approach, Dymola users even do not notice anything: Making "update" will install MSL on their computer so that tables work without further action.

I do not understand the issue. If the trunk version does not work for you for some reason, please explain and maybe it is possible to fix. In the worstcase, just wait for the tagged version on Wednesday.

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#46;se on 20 May 2013 12:14 UTC**
I would have done it the other way around. The version with silly hacks for old tools should be in a branch. We release and work against trunk versions of msl. You will need a maintenance branch with and without the workaround.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 17 Jun 2013 10:01 UTC**
This has been fixed with #1155
